Add a section on interoperability requirements. (#104)
mlswg/mls-architecture: draft-ietf-mls-architecture.md

Upon joining the system, each client stores its initial cryptographic
key material with the Delivery Service. This key material, called a
KeyPackage, advertises the functional abilities of the client such as

> supported protocol versions, extensions, and the following

Reviewer: ```suggestion
supported protocol versions, supported extensions, and the following
```

to preventing downgrade attacks, the members of a group can also observe when it
is safe to upgrade to a new ciphersuite or protocol version.

> # Interoperability Requirements

Reviewer: It seems like these requirements apply not just for cases where two deployments need to interop, but even within a single deployment, in the sense that these are all choices that the deployment has to make.  Maybe something like:

```
# Operational Requirements

MLS is a security layer that needs to be integrated with an application. An
fully-functional deployment of MLS will have to make a number of decisions about
how MLS is configured and operated.  Deployments that wish to interoperate will
need to make compatible decisions. This section lists all of the dependencies of
an MLS deployment that are external to the protocol specification, but would
still need to be aligned within a given MLS deployment, or for two deployments
to potentially interoperate. 
```

> # Interoperability Requirements

Reviewer: It seems like these requirements apply not just for cases where two deployments need to interop, but even within a single deployment, in the sense that these are all choices that the deployment has to make.  Maybe something like:

```
# Operational Requirements

MLS is a security layer that needs to be integrated with an application. An
fully-functional deployment of MLS will have to make a number of decisions about
how MLS is configured and operated.  Deployments that wish to interoperate will
need to make compatible decisions. This section lists all of the dependencies of
an MLS deployment that are external to the protocol specification, but would
still need to be aligned within a given MLS deployment, or for two deployments
to potentially interoperate. 
```

A fully functional deployment of MLS requires a number of infrastructure
services to be available, as well as several policies and design trade-offs that
are decided by the application. This section lists all of the dependencies of an
MLS deployment that are external to the protocol specification, but would still
need to be aligned for two deployments to potentially interoperate.

MLS has the built-in ability to negotiate protocol versions, ciphersuites,

> extensions, and credential types. For two deployments to interoperate, they must

Reviewer: `Capabilities` also has proposal types.

Reviewer: It might be worth mentioning that `required_capabilities` can help ensure that things fail cleanly when required things aren't supported.

> extensions, and credential types. For two deployments to interoperate, they must

Reviewer: `Capabilities` also has proposal types.

have overlapping support in each of these categories.

MLS relies on the following network services. These network services would

> either need to be the same, or compatible at the wire-format level, in order for

Reviewer: ```suggestion
need to be compatible in order for
```

> either need to be the same, or compatible at the wire-format level, in order for

Reviewer: ```suggestion
need to be compatible in order for
```

two different deployments based on them to interoperate.

- An **Authentication Service**, described fully in {{authentication-service}},
  defines the types of credentials which may be used in a deployment and
  provides methods for:
  1. Issuing new credentials,

>   2. Validating existing credentials against a reference identifier, and

Reviewer: ```suggestion
  2. Validating a credential against a reference identifier, and
```
```suggestion
  2. Validating existing credentials against a reference identifier, and
```

>   2. Validating existing credentials against a reference identifier, and

Reviewer: ```suggestion
  2. Validating a credential against a reference identifier, and
```

  3. Validating whether or not two credentials represent the same user.
- A **Delivery Service**, described fully in {{delivery-service}}, provides
  methods for:
  1. Delivering messages sent to a group to all members in the group.
  2. Delivering Welcome messages to new members of a group.
  3. Downloading KeyPackages for specific clients, and uploading new KeyPackages
     for a user's own clients.
- Additional services may or may not be required depending on the application
  design:
  - If assisted joining is desired (meaning that the ratchet tree is not
    provided in Welcome messages), there must be a method to download the
    ratchet tree corresponding to a group.
  - If assisted joining is desired and the Delivery Service is not able to
    compute the ratchet tree itself (because some proposals or commits are sent
    encrypted), there must be a method for group members to publish the updated

>     ratchet tree after each commit.

Reviewer: You could summarize this in one bullet with something like "The tree can be either computed by the DS or uploaded by a group member after each Commit."

>     ratchet tree after each commit.

Reviewer: You could summarize this in one bullet with something like "The tree can be either computed by the DS or uploaded by a group member after each Commit."

    sync with the state of the group.
  - If an application chooses not to allow assisted or external joining, it may
    instead provide a method for external users to solicit group members (or a
    designated service) to add them to a group.
  - If the application uses external PSKs, or uses resumption PSKs that all
    members of a group may not have access to, there must be a method for
    distributing these PSKs to group members.
  - If an application wishes to detect and possibly discipline members that send
    malformed commits with the intention of corrupting a group's state, there
    must be a method for reporting and validating malformed commits.

MLS requires the following parameters to be defined, which must be the same for
two implementations to interoperate:

- The maximum total lifetime that is acceptable for a KeyPackage.
- How long to store the resumption secret for past epochs of a group.
- The degree of tolerance that's allowed for out-of-order message delivery:
  - How long to keep unused nonce and key pairs for a sender
  - A maximum number of unused key pairs to keep.
  - A maximum number of steps that clients will move a secret tree ratchet

>     forward in response to a single message before rejecting it.

Reviewer: These don't have to be the same, do they?  For example, if the DS doesn't allow out-of-order delivery, then you would never detect a difference.  Maybe just soften the "must be the same", e.g., to "must be compatible"

>     forward in response to a single message before rejecting it.

Reviewer: These don't have to be the same, do they?  For example, if the DS doesn't allow out-of-order delivery, then you would never detect a difference.  Maybe just soften the "must be the same", e.g., to "must be compatible"

Reviewer 2: I'll let you submit a subsequent PR if you think an additional action is needed.

MLS requires the following policies to be defined, which restrict the set of
acceptable behavior in a group. These policies must be consistent between
deployments for them to interoperate:

- A policy on when to send proposals and commits in plaintext instead of
  encrypted.
- A policy for which proposals are valid to have in a commit, including but not
  limited to:
  - When a member is allowed to add or remove other members of the group.
  - When, and under what circumstances, a reinitialization proposal is allowed.
  - When proposals from external senders are allowed.
  - When external joiners are allowed.
- A policy for when two credentials represent the same client. Note that many
  credentials may be issued authenticating the same identity but for different
  signature keys, because each credential corresponds to a different device
  (client) owned by the same application user. However, one device may control
  many signature keys but should still only be considered a single client.
- A policy on how long to allow a member to stay in a group without updating its
  leaf keys, and whether there should be any attempt to resolve a stale leaf key

>   before removing the member.

Reviewer: What do you mean "resolve a stale leaf key"?  Like prompt the client to update?  I would just remove this last clause, or amend to something like "... and what actions should be taken with regard to stale leaf keys".

>   before removing the member.

Reviewer: What do you mean "resolve a stale leaf key"?  Like prompt the client to update?  I would just remove this last clause, or amend to something like "... and what actions should be taken with regard to stale leaf keys".
